### PR TITLE
Add gzip support

### PR DESCRIPTION
**Pull request by Athishpranav2003**
Added support for gzip compression while exposing metrics. 
Tested locally with prometheus server
image

#219

### CODE REVIEW

File: lib/fluent/plugin/in_prometheus.rb
Comment (Lusitaniae) [inline lib/fluent/plugin/in_prometheus.rb:62]: what do you mean identity?

Comment (Athishpranav2003) [reply lib/fluent/plugin/in_prometheus.rb:62]: https://devdoc.net/web/developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding.html
Basically its same as no compression

File: README.md
Comment (Lusitaniae) [inline README.md:66]: usually you'd read `Content-Accept` header from the request to determine how to encode the response

**Comment by Lusitaniae** [comment]
amazing turnaround, much appreciated for your efforts @Athishpranav2003!

File: README.md
Comment (Athishpranav2003) [reply README.md:66]: This is true
But in our case Prometheus server accepts both gzip and identity (it's not specific) so the request header has both in accept encoding. We cannot determine whether we only need compressed or identity. That's why I configured it in server side

**Comment by Athishpranav2003** [comment]
Thanks @Lusitaniae 
One help
The signoff check is not passing for me from first.
Initially I had dummy author creds in local git so it messed it. I reverted to original stuff and still not passing the check. Can you help me with that alone. 
Also who has write access to merge this?

**Comment by Athishpranav2003** [comment]
@ashie can you please check this

**Comment by ashie** [comment]
@Athishpranav2003  Thanks for your contribution!
Could you follow DCO as @Lusitaniae mentioned?
We require `Signed-off-by:` field in each commit message.

> Looks like you just need to follow these steps:
> 
> > In your local branch, run: git rebase HEAD~2 --signoff
> > Force push your changes to overwrite the branch: git push --force-with-lease origin add-gzip-support
> 
> https://github.com/fluent/fluent-plugin-prometheus/pull/223/checks?check_run_id=28060535686

File: fluent-plugin-prometheus.gemspec
Comment (ashie) [inline fluent-plugin-prometheus.gemspec:3]: Could you remove this version change from this pull request?
It will be controlled by us.
In addition, this pull request adding a new feature, so the minor version should be incremented instead of the patch version.

Comment (Athishpranav2003) [reply fluent-plugin-prometheus.gemspec:3]: Yah sure
I will revert this file

**Comment by Athishpranav2003** [comment]
@ashie I have made the changes. In addition i have added zlib as dependency to avoid break incase some system misses it(highly unlikely).

**Comment by ashie** [comment]
I'll fix CI for Ruby 2.7 in another pull request, so please leave it.

**Comment by Athishpranav2003** [comment]
@ashie do you need help in testing this locally before merging

**Comment by ashie** [comment]
It looks good, I'll merge this soon after checking if we missed anything in relation to the surrounding code.

File: lib/fluent/plugin/in_prometheus.rb
Comment (ashie) [inline lib/fluent/plugin/in_prometheus.rb:37]:
```suggestion
    config_param :content_encoding, :enum, list: [:identify, :gzip], default: :identify
```

It would be better to use the config value type :enum instead of :string to specify fixed choices, so that following check by plugin isn't needed.

File: lib/fluent/plugin/in_prometheus.rb
Comment (ashie) [inline lib/fluent/plugin/in_prometheus.rb:62]:
```suggestion
      raise "Invalid content encoding for the exposed metrics endpoint" unless @content_encoding == "identity" || @content_encoding == "gzip"
```

`@content_encoding="identity"` assigns `identify` to `@content_encoding` and always returns `true`.
So that `@content_encoding` is always `identify` in actual.

Although above suggestion is one of solution, it would be better to change the type of `content_encoding` to `:enum` and remove this line to make it simpler.

File: lib/fluent/plugin/in_prometheus.rb
Comment (ashie) [inline lib/fluent/plugin/in_prometheus.rb:195]:
```suggestion
      when :gzip
```

(If you change the type of it to `:enum`.)

File: lib/fluent/plugin/in_prometheus.rb
Comment (ashie) [inline lib/fluent/plugin/in_prometheus.rb:199]: Ditto

File: lib/fluent/plugin/in_prometheus.rb
Comment (ashie) [inline lib/fluent/plugin/in_prometheus.rb:202]:
```suggestion
      [200, { 'Content-Type' => ::Prometheus::Client::Formats::Text::CONTENT_TYPE, 'Content-Encoding' => @content_encoding.to_s }, body]
```

(Ditto)

File: lib/fluent/plugin/in_prometheus.rb
Comment (ashie) [inline lib/fluent/plugin/in_prometheus.rb:217]: Ditto.

File: lib/fluent/plugin/in_prometheus.rb
Comment (ashie) [inline lib/fluent/plugin/in_prometheus.rb:221]: Ditto.

**Review by ashie — CHANGES_REQUESTED** [review]
I have made some comments about some things I would like to see fixed.
In addition, it would be better to add tests for this feature if possible.

**Comment by Athishpranav2003** [comment]
Ok
I will take a look at the comments.

Regarding UTs there is no UTs written for the plugin.
So the UTs have go in a seperate PR if we want a Minitest kinda UTs. For now I was manually testing the changes.

File: lib/fluent/plugin/in_prometheus.rb
Comment (Athishpranav2003) [reply lib/fluent/plugin/in_prometheus.rb:37]: One doubt here
Does the list have to be list of symbols?

Comment (ashie) [reply lib/fluent/plugin/in_prometheus.rb:37]: Yes, list items have to be symbol:
https://github.com/fluent/fluentd/blob/64595e2ac291a3e1d2d02feefdc58e623b3abb02/lib/fluent/config/types.rb#L126

**Comment by Athishpranav2003** [comment]
@ashie i have addressed your comments and also refactored some small part of my code

**Comment by Athishpranav2003** [comment]
@ashie i am working on UTs part. I can get that also done quick.

**Comment by Athishpranav2003** [comment]
@ashie i have added UTs now
please take a look at it when u are free

File: spec/fluent/plugin/in_prometheus_spec.rb
Comment (ashie) [inline spec/fluent/plugin/in_prometheus_spec.rb:229]:
```suggestion
            req['accept-encoding'] = nil
            res = http.request(req)
```

`Net::HTTP::Response` deflates the compressed data automatically by default, so that it's difficult to recognize whether this feature is really working as expected or not.
To distinguish whether the contents is compressed or not, I recommend to unset `Accept-Encoding` to disable automatic deflating.

File: spec/fluent/plugin/in_prometheus_spec.rb
Comment (ashie) [inline spec/fluent/plugin/in_prometheus_spec.rb:247]:
```suggestion
            req['accept-encoding'] = nil
            res = http.request(req)
            body = Zlib::GzipReader.new(StringIO.new(res.body)).read
            expect(body).to include("test Testing metrics")
```

Ditto.

**Review by ashie — CHANGES_REQUESTED** [review]

**Comment by Athishpranav2003** [comment]
@ashie addressed the comments

File: fluent-plugin-prometheus.gemspec
Comment (ashie) [inline fluent-plugin-prometheus.gemspec:18]: Sorry, one more thing.
This dependency isn't needed because it's a default gem (bundled in Ruby and cannot remove), and this PR doesn't use specific features in zlib >= 3.0.0.
In fact fluentd's bundled plugins also use it without such dependency.

Comment (Athishpranav2003) [reply fluent-plugin-prometheus.gemspec:18]: OK 
i will remove it
anything else?

Comment (ashie) [reply fluent-plugin-prometheus.gemspec:18]: No, I'm going to merge this after it's removed.

Strictly speaking it would be better to check `Accept-Encoding` then response uncompressed data if it doesn't support gzip. However, since behaviour of client is known and controllable on this case as you mentioned in https://github.com/fluent/fluent-plugin-prometheus/pull/223#discussion_r1695554372, there is no need to go that far.

**Review by ashie — APPROVED** [review]
LGTM

**Comment by ashie** [comment]
I've removed wrongly added fluent-plugin-prometheus-2.1.0.gem in your commit.

**Comment by ashie** [comment]
Thanks!

**Pull request merged**
